pkg/bootstrap/docker/dockerhelper/client.go

```diff
@@ -0,0 +1,333 @@
+package dockerhelper
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/engine-api/client"
+	"github.com/docker/engine-api/types"
+	"github.com/docker/engine-api/types/container"
+	"github.com/docker/engine-api/types/network"
+)
+
+const (
+	// defaultTimeout is the default timeout of short running docker operations.
+	defaultTimeout = 2 * time.Minute
+)
+
+// NewClient creates an instance of the client Interface, given a docker engine
+// client
+func NewClient(endpoint string, client *client.Client) Interface {
+	return &dockerClient{
+		endpoint: endpoint,
+		client:   client,
+	}
+}
+
+type dockerClient struct {
+	endpoint string
+	client   *client.Client
+}
+
+type containerClient struct {
+	name   string
+	client *client.Client
+}
+
+type imageClient struct {
+	name   string
+	client *client.Client
+}
+
+type execClient struct {
+	name   string
+	client *client.Client
+}
+
+type operationTimeout struct {
+	err error
+}
+
+type apiResult interface{}
+
+type apiCallFunc func(ctx context.Context) (apiResult, error)
+
+func (e operationTimeout) Error() string {
+	return fmt.Sprintf("operation timeout: %v", e.err)
+}
+
+func (c *dockerClient) Endpoint() string {
+	return c.endpoint
+}
+
+func (c *dockerClient) Version() (*types.Version, error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	version, err := c.client.ServerVersion(ctx)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return nil, ctxErr
+	}
+	return &version, err
+}
+
+func (c *dockerClient) Info() (*types.Info, error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	info, err := c.client.Info(ctx)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return nil, ctxErr
+	}
+	return &info, err
+}
+
+func (c *dockerClient) ContainerNames() ([]string, error) {
```

Comment:
would prefer this to be ContainerList - move the additional smarts to the caller - there's only one.